File: web/src/lib/geometry/mom_capacitor.ts
import type { Polygon, LayerMap, LayerName, MomCapacitorParams } from './types';
import type { ConductorNetwork, ConductorNode, Port, GeometryResult } from './network';
import { viaGrid } from './utils';

/**
 * Build a MOM (Metal-Oxide-Metal) interdigitated finger capacitor.
 *
 * Features:
 *   - Edge shielding: outermost fingers on both sides are the same polarity (P)
 *   - Half-pitch offset between adjacent metal layers for lateral + vertical coupling
 *   - Per-finger via arrays connecting same-polarity fingers across layers
 *   - Bus bars sized to fully enclose finger bases
 */
export function buildMomCapacitor(params: MomCapacitorParams): GeometryResult {
	const { nFingers, fingerLength, fingerWidth, fingerSpacing,
		busWidth, nLayers, via_spacing, via_width, via_in_metal } = params;

	const layers: LayerMap = {};
	const pitch = fingerWidth + fingerSpacing;

	// Ensure odd finger count for edge shielding (both edges = P)
	const nf = nFingers % 2 === 0 ? nFingers + 1 : nFingers;

	// Layer assignments per metal level
	const metalLayers: LayerName[] = ['windings', 'windings_m2', 'crossings_m1'];
	const viaLayers: LayerName[] = ['vias1', 'vias2'];

	// Compute total width including half-pitch extension for offset layers
	const baseWidth = nf * fingerWidth + (nf - 1) * fingerSpacing;
	const offsetExtra = pitch / 2;
	const totalWidth = baseWidth + offsetExtra; // wide enough for offset layers
	const totalHeight = fingerLength + 2 * busWidth;

	// Center at origin
	const x0 = -totalWidth / 2;
	const y0 = -totalHeight / 2;

	for (let layer = 0; layer < nLayers; layer++) {
		const renderLayer = metalLayers[layer];
		const polys: Polygon[] = [];

		// Half-pitch offset on odd layers
		const offset = layer % 2 === 0 ? 0 : pitch / 2;

		// Finger x-start (centered within the total width)
		const layerWidth = nf * fingerWidth + (nf - 1) * fingerSpacing;
		const layerX0 = -layerWidth / 2 + offset;

		// Generate fingers
		for (let i = 0; i < nf; i++) {
			const isPositive = i % 2 === 0; // P at edges (i=0, i=nf-1 both even for odd nf)
			const fx = layerX0 + i * pitch;

			// Skip fingers that fall outside the total width
			if (fx + fingerWidth > totalWidth / 2 + 0.001) continue;
			if (fx < -totalWidth / 2 - 0.001) continue;

			let fy: number, fh: number;
			if (isPositive) {
				// P finger: extends up from bottom bus, gap at top
				fy = y0 + busWidth;
				fh = fingerLength - fingerSpacing;
			} else {
				// N finger: extends down from top bus, gap at bottom
				fy = y0 + busWidth + fingerSpacing;
				fh = fingerLength - fingerSpacing;
			}

			polys.push({
				x: [fx, fx + fingerWidth, fx + fingerWidth, fx],
				y: [fy, fy, fy + fh, fy + fh],
			});
		}

		// P bus bar (bottom) — full total width to cover offset fingers
		polys.push({
			x: [x0, x0 + totalWidth, x0 + totalWidth, x0],
			y: [y0, y0, y0 + busWidth, y0 + busWidth],
		});

		// N bus bar (top) — full total width
		polys.push({
			x: [x0, x0 + totalWidth, x0 + totalWidth, x0],
			y: [y0 + busWidth + fingerLength, y0 + busWidth + fingerLength, y0 + totalHeight, y0 + totalHeight],
		});

		layers[renderLayer] = [...(layers[renderLayer] || []), ...polys];

		// Via arrays between this layer and the one above
		if (layer > 0) {
			const viaLayer = viaLayers[layer - 1];
			let viaPolys: Polygon[] = [];

			// Per-finger vias where fingers overlap between layers
			// Fingers on adjacent layers are offset by half-pitch, so each finger
			// overlaps with the adjacent layer's bus bar at its base.
			// Place vias on the bus bars (full overlap guaranteed)
			const busViaW = totalWidth - 2 * via_in_metal;
			const busViaH = busWidth - 2 * via_in_metal;

			if (busViaW > 0 && busViaH > 0) {
				// P bus vias (bottom)
				viaPolys = viaPolys.concat(
					viaGrid(0, y0 + busWidth / 2, busViaW, busViaH, via_spacing, via_width)
				);
				// N bus vias (top)
				viaPolys = viaPolys.concat(
					viaGrid(0, y0 + busWidth + fingerLength + busWidth / 2, busViaW, busViaH, via_spacing, via_width)
				);
			}

			// Per-finger vias in the overlapping finger region
			const prevOffset = (layer - 1) % 2 === 0 ? 0 : pitch / 2;
			const currOffset = layer % 2 === 0 ? 0 : pitch / 2;
			const prevLayerX0 = -layerWidth / 2 + prevOffset;
			const currLayerX0 = -layerWidth / 2 + currOffset;

			for (let i = 0; i < nf; i++) {
				const prevFx = prevLayerX0 + i * pitch;
				const prevIsP = i % 2 === 0;

				for (let j = 0; j < nf; j++) {
					const currFx = currLayerX0 + j * pitch;
					const currIsP = j % 2 === 0;

					// Only connect same polarity
					if (prevIsP !== currIsP) continue;

					// Check overlap
					const overlapL = Math.max(prevFx, currFx);
					const overlapR = Math.min(prevFx + fingerWidth, currFx + fingerWidth);
					const overlapW = overlapR - overlapL;
					if (overlapW < via_width + 2 * via_in_metal) continue;

					// Finger overlap region in Y
					const fyCx = (overlapL + overlapR) / 2;
					const fingerTop = y0 + busWidth + fingerLength - fingerSpacing;
					const fingerBot = y0 + busWidth + fingerSpacing;
					const fyCy = (fingerBot + fingerTop) / 2;
					const fingerViaH = (fingerTop - fingerBot) - 2 * via_in_metal;
					const fingerViaW = overlapW - 2 * via_in_metal;

					if (fingerViaH > 0 && fingerViaW > 0) {
						viaPolys = viaPolys.concat(
							viaGrid(fyCx, fyCy, fingerViaW, fingerViaH, via_spacing, via_width)
						);
					}
				}
			}

			layers[viaLayer] = [...(layers[viaLayer] || []), ...viaPolys];
		}
	}

	// Network with port markers
	const nodes: ConductorNode[] = [
		{ id: 'p_plus', x: x0, y: y0 + busWidth / 2, layerId: 'm3' },
		{ id: 'p_minus', x: x0 + totalWidth, y: y0 + busWidth + fingerLength + busWidth / 2, layerId: 'm3' },
	];
	const ports: Port[] = [
		{ name: 'P+', node: 'p_plus' },
		{ name: 'P-', node: 'p_minus' },
	];

	const network: ConductorNetwork = { nodes, segments: [], vias: [], ports };

	return { network, layers };
}

export function isMomCapacitorValid(params: MomCapacitorParams): boolean {
	const { nFingers, fingerLength, fingerWidth, fingerSpacing, busWidth, nLayers,
		via_width, via_spacing, via_in_metal } = params;

	if (nFingers < 2 || fingerLength <= 0 || fingerWidth <= 0 || fingerSpacing <= 0) return false;
	if (busWidth <= 0 || nLayers < 1 || nLayers > 3) return false;
	if (via_width <= 0 || via_spacing <= 0 || via_in_metal < 0) return false;

	// Fingers must be longer than the gap to actually interdigitate
	if (fingerLength <= 2 * fingerSpacing) return false;

	// Bus bar must be wide enough for at least one via row when multi-layer
	if (nLayers > 1 && busWidth < 2 * via_in_metal + via_width) return false;

	return true;
}

File: rapidpassives/geometry/mom_capacitor.py
"""MOM interdigitated capacitor — Python port of web/src/lib/geometry/mom_capacitor.ts."""
from __future__ import annotations

from .primitives import Poly, via_grid
from .result import GeometryResult, Port

_METAL_LAYERS = ["windings", "windings_m2", "crossings_m1"]
_VIA_LAYERS = ["vias1", "vias2"]


def build_mom_capacitor(params: dict) -> GeometryResult:
    n_fingers = params["nFingers"]; finger_length = params["fingerLength"]
    finger_width = params["fingerWidth"]; finger_spacing = params["fingerSpacing"]
    bus_width = params["busWidth"]; n_layers = params["nLayers"]
    via_spacing = params["via_spacing"]; via_width = params["via_width"]; via_in_metal = params["via_in_metal"]

    layers: dict[str, list[Poly]] = {}
    pitch = finger_width + finger_spacing
    nf = n_fingers + 1 if n_fingers % 2 == 0 else n_fingers

    base_width = nf * finger_width + (nf - 1) * finger_spacing
    offset_extra = pitch / 2
    total_width = base_width + offset_extra
    total_height = finger_length + 2 * bus_width
    x0 = -total_width / 2
    y0 = -total_height / 2

    for layer in range(n_layers):
        render_layer = _METAL_LAYERS[layer]
        polys: list[Poly] = []
        offset = 0 if layer % 2 == 0 else pitch / 2
        layer_width = nf * finger_width + (nf - 1) * finger_spacing
        layer_x0 = -layer_width / 2 + offset

        for i in range(nf):
            is_positive = i % 2 == 0
            fx = layer_x0 + i * pitch
            if fx + finger_width > total_width / 2 + 0.001:
                continue
            if fx < -total_width / 2 - 0.001:
                continue
            if is_positive:
                fy = y0 + bus_width; fh = finger_length - finger_spacing
            else:
                fy = y0 + bus_width + finger_spacing; fh = finger_length - finger_spacing
            polys.append([(fx, fy), (fx + finger_width, fy), (fx + finger_width, fy + fh), (fx, fy + fh)])

        # Bus bars (full width).
        polys.append([(x0, y0), (x0 + total_width, y0), (x0 + total_width, y0 + bus_width), (x0, y0 + bus_width)])
        ty = y0 + bus_width + finger_length
        polys.append([(x0, ty), (x0 + total_width, ty), (x0 + total_width, y0 + total_height), (x0, y0 + total_height)])

        layers[render_layer] = [*layers.get(render_layer, []), *polys]

        if layer > 0:
            via_layer = _VIA_LAYERS[layer - 1]
            via_polys: list[Poly] = []
            bus_via_w = total_width - 2 * via_in_metal
            bus_via_h = bus_width - 2 * via_in_metal
            if bus_via_w > 0 and bus_via_h > 0:
                via_polys += via_grid(0, y0 + bus_width / 2, bus_via_w, bus_via_h, via_spacing, via_width)
                via_polys += via_grid(0, y0 + bus_width + finger_length + bus_width / 2, bus_via_w, bus_via_h, via_spacing, via_width)

            prev_offset = 0 if (layer - 1) % 2 == 0 else pitch / 2
            curr_offset = 0 if layer % 2 == 0 else pitch / 2
            prev_x0 = -layer_width / 2 + prev_offset
            curr_x0 = -layer_width / 2 + curr_offset
            for i in range(nf):
                prev_fx = prev_x0 + i * pitch
                prev_is_p = i % 2 == 0
                for j in range(nf):
                    curr_fx = curr_x0 + j * pitch
                    curr_is_p = j % 2 == 0
                    if prev_is_p != curr_is_p:
                        continue
                    overlap_l = max(prev_fx, curr_fx)
                    overlap_r = min(prev_fx + finger_width, curr_fx + finger_width)
                    overlap_w = overlap_r - overlap_l
                    if overlap_w < via_width + 2 * via_in_metal:
                        continue
                    fy_cx = (overlap_l + overlap_r) / 2
                    finger_top = y0 + bus_width + finger_length - finger_spacing
                    finger_bot = y0 + bus_width + finger_spacing
                    fy_cy = (finger_bot + finger_top) / 2
                    finger_via_h = (finger_top - finger_bot) - 2 * via_in_metal
                    finger_via_w = overlap_w - 2 * via_in_metal
                    if finger_via_h > 0 and finger_via_w > 0:
                        via_polys += via_grid(fy_cx, fy_cy, finger_via_w, finger_via_h, via_spacing, via_width)

            layers[via_layer] = [*layers.get(via_layer, []), *via_polys]

    ports = [
        Port("P+", x0, y0 + bus_width / 2, "windings"),
        Port("P-", x0 + total_width, y0 + bus_width + finger_length + bus_width / 2, "windings"),
    ]
    return GeometryResult(layers, ports)
